googledataprocauthenticator/google.py

```diff
@@ -0,0 +1,319 @@
+        ValueError: If the parameters are invalid.
+    """
+    client = dataproc_v1beta2.ClusterControllerClient(credentials=credentials,
+                       client_options={
+                            'api_endpoint': '{}-dataproc.googleapis.com:443'.format(region)
+                        }
+                    )
+    try:
+        print(get_cluster_pool(project_id, region, client))
+        # if they didn't enter cluster name then we get a cluster from cluster pool 
+        cluster_pool = None
+        if cluster_name is '':
+            #pop random cluster from cluster pool 
+            cluster_pool = get_cluster_pool(project_id, region, client)
+            cluster_name = cluster_pool.copy().pop()
+        response = client.get_cluster(project_id, region, cluster_name)
+        url = ((response.config.endpoint_config).http_ports).popitem()[1]
+        parsed_uri = urllib3.util.parse_url(url)
+        endpoint_address = '{uri.scheme}://{uri.netloc}/'.format(uri=parsed_uri) + 'gateway/default/livy/v1'
+        ipython_display = IpythonDisplay()
+        
+        if cluster_pool is not None:
+            ipython_display.writeln("Used {} cluster from cluster_pool: {}".format(cluster_name, cluster_pool))
+        else:
+            ipython_display.writeln("Used {} cluster".format(cluster_name))
+        return endpoint_address
+    except:
+        raise
+
+def application_default_credentials_configured(): 
+    """Checks if google application-default credentials are configured"""
+    callable_request = google.auth.transport.requests.Request()
+    try:
+        credentials, project = google.auth.default(scopes=['https://www.googleapis.com/auth/cloud-platform','https://www.googleapis.com/auth/userinfo.email'])
+    except:
+        pass
+        return False
+    return not(credentials is None)
+
+
+class GoogleAuth(Authenticator):
+    """Custom Authenticator to use Google OAuth with SparkMagic."""
+
+    def __init__(self, parsed_attributes=None):
+        self.callable_request = google.auth.transport.requests.Request()
+        self.credentialed_accounts, active_user_account = list_credentialed_accounts()
+        self.scopes = ['https://www.googleapis.com/auth/cloud-platform', 'https://www.googleapis.com/auth/userinfo.email']
+        self.active_credentials = None
+        self.default_credentials_configured = application_default_credentials_configured()
+        
+        account_dict = list_accounts_pairs(self.credentialed_accounts, self.default_credentials_configured)
+        if parsed_attributes is not None:
+            if parsed_attributes.account in account_dict:
+                self.active_credentials = parsed_attributes.account
+            else: 
+                new_exc = BadUserConfigurationException(
+                "{} is not a credentialed account. Run `gcloud auth login` in your command line "\
```

Comment:
Nit: I think the spacing is a little bit off here.

BTW, assuming your library is Python 3-only, you can use f-strings rather than ''.format(): https://realpython.com/python-f-strings/